### PR TITLE
GEODE-7497: Check CQs prior to change authorizer

### PR DESCRIPTION
**Pull request by jujoramos**
The QueryConfigurationService now requires an extra flag to determine,
in the presence of continuous queries, whether to update the configured
MethodInvocationAuthorizer and invalidate the CQ's internal cache or
throw an exception and abort the update.

- Fixed minor warnings.
- Added unit and distributed tests.
- Updated docs for alter query-service command.
- Added 'forceUpdate' flag to 'AlterQueryServiceFunction' and
  'AlterQueryServiceCommand'.
- Fixed 'AlterQueryServiceCommand' to always require the
  'method-authorizer' parameter.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by jujoramos** [comment]
@DonalEvans @BenjaminPerryRoss

File: geode-core/src/distributedTest/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandWithSecurityDUnitTest.java [hunk 1]
```diff
@@ -61,81 +61,119 @@
...
+  private void verifyCurrentAuthorizerClass(Class authorizerClass) {
+    servers.forEach(server -> server.invoke(() -> {
+      MethodInvocationAuthorizer methodAuthorizer =
+          Objects.requireNonNull(ClusterStartupRule.getCache())
+              .getService(QueryConfigurationService.class).getMethodAuthorizer();
+      assertThat(methodAuthorizer.getClass()).isEqualTo(authorizerClass);
+    }));
+  }
+
+  private String getFilePath() throws IOException {
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandWithSecurityDUnitTest.java]: Since this method now always gets the file path for the `TestMethodAuthorizer.txt` file, it might be better to rename it something like "getTestMethodAuthorizerFilePath" to make it clearer what it's doing when called.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImplTest.java [hunk 2]
```diff
@@ -43,49 +54,70 @@
...
+  private void setAllowUntrustedMethodInvocationSystemProperty() {
+    System.setProperty(ALLOW_UNTRUSTED_METHOD_INVOCATION_SYSTEM_PROPERTY, "true");
   }
 
   @Test
   public void initThrowsExceptionWhenCacheIsNull() {
     assertThatThrownBy(() -> configService.init(null))
-        .isInstanceOf(IllegalArgumentException.class);
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage("Cache must not be null");
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImplTest.java]: If this verifcation of the message is being done, then the string used to generate the message in `QueryConfigurationServiceImpl` should be extracted to a constant and referenced here.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java [hunk 3]
```diff
@@ -50,163 +50,191 @@
...
+    }
+
+    return commandBuilder.toString();
   }
 
   @Test
-  public void commandReturnsCorrectMessageWithNoOptions() {
-    gfsh.executeAndAssertThat(command, COMMAND_NAME).statusIsError()
-        .containsOutput(NO_ARGUMENTS_MESSAGE);
+  public void commandParsesForceUpdateCorrectly() {
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java]: Is this test needed here? It seems to be testing the functionality of the gfsh parser rather than the contents of the `AltherQueryServiceCommand` class.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java [hunk 4]
```diff
@@ -31,78 +29,70 @@
...
       operation = ResourcePermission.Operation.MANAGE)
   public ResultModel execute(
-      @CliOption(key = METHOD_AUTHORIZER_NAME,
-          help = METHOD_AUTHORIZER_NAME_HELP) String methodAuthorizerName,
-      @CliOption(key = AUTHORIZER_PARAMETERS,
-          help = AUTHORIZER_PARAMETERS_HELP,
-          optionContext = "splittingRegex=" + SPLITTING_REGEX) String[] authorizerParameters) {
-    ResultModel result;
+      @CliOption(key = AUTHORIZER_NAME, help = METHOD_AUTHORIZER_NAME_HELP,
+          mandatory = true) String methodAuthorizerName,
```
[the remark this answers is not in the document]
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java]: Thanks for the explanation, it makes total sense!. I'd leave as it is right now, though, and make it optional once we actually need it.

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationService.java [hunk 5]
```diff
@@ -24,12 +24,41 @@
...
+   */
   MethodInvocationAuthorizer getMethodAuthorizer();
 
-  void updateMethodAuthorizer(Cache cache, QueryMethodAuthorizerCreation creation)
-      throws ClassNotFoundException;
-
-  void updateMethodAuthorizer(Cache cache, String className, Set<String> parameters)
-      throws ClassNotFoundException, QueryConfigurationServiceException;
+  /**
+   * Updates the configured {@link MethodInvocationAuthorizer}.
```
[the remark this answers is not in the document]
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationService.java]: Good point, thanks.
I've updated the javadoc and also deleted the `forceUpdate` parameter from this method, it should always be `true` whenever we're creating the cache from scratch.

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationService.java [hunk 6]
```diff
@@ -24,12 +24,41 @@
...
+   *
+   * @param cache the cache on which the {@link MethodInvocationAuthorizer} instance will be
+   *        configured.
+   * @param forceUpdate {@code true} to apply the configuration change even when there are
+   *        continuous queries running, {@code false} otherwise.
+   * @param className the fully qualified name of the class that will be created and configured as
+   *        the new {@link MethodInvocationAuthorizer}.
+   * @param parameters the list of parameters that will be used to initialize the new
```
[the remark this answers is not in the document]
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationService.java]: Done!.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java [hunk 7]
```diff
@@ -31,78 +29,70 @@
...
-      "A semicolon separated list of all parameter values for the specified method authorizer. Requires '--"
-          + METHOD_AUTHORIZER_NAME + "' option to be used";
-  static final String PARAMETERS_WITHOUT_AUTHORIZER_MESSAGE = "The '--" + AUTHORIZER_PARAMETERS
-      + "' option requires '--" + METHOD_AUTHORIZER_NAME + "' to be specified";
-  static final String NO_ARGUMENTS_MESSAGE =
-      "No arguments were provided. No changes have been applied.";
+      "A semicolon separated list of all parameter values for the specified method authorizer.";
+  static final String FORCE_UPDATE = "force-update";
+  private static final String FORCE_PARAMETER_HELP =
+      "Flag to indicate whether to forcibly update the currently configured authorizer, even when there are continuous queries registered (use with caution)";
```
[the remark this answers is not in the document]
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java]: I've changed the parameter name, thanks for pointing that out.
I'm not sure about adding the part about invalidating CQs, though, it might be too much for actually being "helpful" to a user (maybe they don't even have CQs at all, adding confusion instead of help). Also, we only invalidate the internal cache used by CQs (internal mechanism totally unknown to our users), I wouldn't know how to phrase this warning without going into much detail and confuse even more the operator...
The user always has the option to go and check the documentation if needed, and the `use with caution` part should be "enough" to make them think twice before using the flag.
Please let me know your thoughts around this.

File: geode-core/src/distributedTest/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandWithSecurityDUnitTest.java (hunk 1 above)
Comment (jujoramos) [reply geode-core/src/distributedTest/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandWithSecurityDUnitTest.java]: Sounds good, done!.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImplTest.java (hunk 2 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImplTest.java]: Done!.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java (hunk 3 above)
Comment (jujoramos) [reply geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java]: I wrote the test **before** actually adding the new parameter `forceUpdate` with its associated constraints (`specifiedDefaultValue = "true", unspecifiedDefaultValue = "false"`), so I'm actually testing that what I did was right.
Long story short, I used TDD for this one and I'd like to keep these tests.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java]: I think you're right that if the cache that's invalidated is entirely internal then we don't need to tell users about it. At that point it's just making things more confusing.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java]: I think you're probably right. We don't even know at this point if we do plan to add other persistence to the `QueryConfigurationService`, so we can cross that bridge when we come to it.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java (hunk 3 above)
Comment (DonalEvans) [reply geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommandTest.java]: That makes sense.

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImpl.java [hunk 8]
```diff
@@ -141,12 +156,18 @@ public void updateMethodAuthorizer(Cache cache, String className, Set<String> pa
...
+
         MethodInvocationAuthorizer tmpAuthorizer =
             (MethodInvocationAuthorizer) userClass.newInstance();
         tmpAuthorizer.initialize(cache, parameters);
         this.authorizer = tmpAuthorizer;
-
       }
+
+      // Invalidate cached CQ results, if any.
+      cqService.getAllCqs().forEach(cqQuery -> {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImpl.java]: Should this be pulled into a method?  It would be self documenting (no comment needed) if the method name was correctly chosen

Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImpl.java]: Done!

**Comment by jujoramos** [comment]
Thanks for reviewing @jhuynh1 
Please find my answers below, inline for an easier follow-up.

> I thought clearing the local cq cache was needed if we updated the cq context with the new authorizer.

The CQ will always pick up the latest configured authorizer, no matter wether the `forceUpdate` flag is set as `true` or `false`, this logic is part of the `reset` method invoked by CQs during the execution. See [`CqServiceImpl.evaluateQuery`](https://github.com/apache/geode/blob/develop/geode-cq/src/main/java/org/apache/geode/cache/query/cq/internal/CqServiceImpl.java#L1604-L1607) and [`ExecutionContext.reset`](https://github.com/apache/geode/blob/develop/geode-core/src/main/java/org/apache/geode/cache/query/internal/ExecutionContext.java#L587-L591).

> If they have the old authorizer, why do we need to clear the cache?

If we don't clear the cache, the CQ won't be evaluated for those keys that are part of the cache, which can result in invoking `onEvent` instead of `onError` for the associated `CqListener`.

---

As an example, let's say the user has the CQ  `SELECT * from /testRegion object WHERE object.getId() > -1` running in the cluster, the configured `MethodInvocationAuthorizer` allows `getId` to be invoked. Some time passes and the CQ internal cache builds up to improve performance.
At that point the user decides to install a new `MethodInvocationAuthorizer` that doesn't allow `getId`. If we don't invalidate the internal CQ cache, the engine will invoke `onError` for events **that are not part of the internal cache** (the CQ will be executed) and `onEvent` for those events that **that are part of the internal cache** (the CQ won't be executed), which is not consistent.
Long story short, by invalidating the internal cache we make sure the CQ will be re-executed for every event  no matter what, and the execution itself will result in error if the newly configured authorizer doesn't allow the execution of methods present in the CQ expression.
Hope it makes sense now, please let me know your thoughts.
Cheers.
